serverside-javascript-best-practices.md

```diff
@@ -320,6 +350,53 @@ createBooking.call(self, product, params, 'upgrades', 'supplier', 'payment', fun
 
 
 
+## Handling heavily sequential async code with promises
+
+* !! Using promises for heavily sequential async code has the downside of hiding the data being passed between functions. We lose the ability to wrap each layer in a closure to selectively pass the correct data betweeen the correct entites. Eveything must flow through the middle to get to the end. This breaks our rules on single-responsibility, but allows for slightly cleaner orchestration:
+* !! Sinon promises do not support sandboxing yet, so you'll have to manually restore everything by hand
+* !! Its impossible to prove that the data flow between promises is correct, only that data moves between them.
+
+```javascript
+createBooking._orchestrate = function(data) {
+  return createBooking._sendBook1Request(data)
+    .then(createBooking._generatePaymentRequest)
+    .then(createBooking._processPayment)
+    .then(createBooking._sendBook2Request)
+    .then(createBooking._generateBookingReference);
+};
+```
+```javascript
+require('sinon-bluebird');
```

Comment:
I'm keen to retain the `.calledWithPromise( )` feature, or rather, I'm happy with any solution that provides the ability to assert what a stub was invoked with. For now putting in a few lines to say this example uses Bluebird to achieve that feels like the best approach.